File: c-interface/include/src/VertexBuffer.cpp
#include "Astral.Canvas/Graphics/VertexBuffer.h"
#include "Graphics/VertexBuffer.hpp"

exportC AstralCanvasVertexDeclaration AstralCanvasVertexBuffer_GetVertexDeclaration(AstralCanvasVertexBuffer ptr)
{
    return (AstralCanvasVertexDeclaration)((AstralCanvas::VertexBuffer *)ptr)->vertexType;
}
exportC usize AstralCanvasVertexBuffer_GetCount(AstralCanvasVertexBuffer ptr)
{
    return ((AstralCanvas::VertexBuffer *)ptr)->vertexCount;
}
exportC AstralCanvasVertexBuffer AstralCanvasVertexBuffer_Create(AstralCanvasVertexDeclaration thisVertexType, usize vertexCount, bool isDynamic, bool canRead)
{
    AstralCanvas::VertexBuffer *buffer = (AstralCanvas::VertexBuffer*)GetDefaultAllocator()->Allocate(sizeof(AstralCanvas::VertexBuffer));
    *buffer = AstralCanvas::VertexBuffer((AstralCanvas::VertexDeclaration*)thisVertexType, vertexCount, isDynamic, canRead);
    return buffer;
}
exportC void AstralCanvasVertexBuffer_Deinit(AstralCanvasVertexBuffer ptr)
{
    ((AstralCanvas::VertexBuffer *)ptr)->deinit();
}
exportC void AstralCanvasVertexBuffer_SetData(AstralCanvasVertexBuffer ptr, void *verticesData, usize verticesCount)
{
    ((AstralCanvas::VertexBuffer *)ptr)->SetData(verticesData, verticesCount);
}

File: bindings/csharp/VertexBuffer.cs
using System;

namespace Astral.Canvas
{
    public class VertexBuffer : IDisposable
    {
        public IntPtr handle;
        public int vertexCount;
        public VertexDeclaration vertexDeclaration;

        public VertexBuffer(VertexDeclaration vertexDeclaration, int vertexCount, bool isDynamic, bool canRead = false)
        {
            handle = AstralCanvas.VertexBuffer_Create(vertexDeclaration.handle, (UIntPtr)vertexCount, isDynamic, canRead);
            this.vertexDeclaration = vertexDeclaration;
            this.vertexCount = vertexCount;
        }
        public unsafe void SetData<T>(T[] vertices) where T : unmanaged
        {
            fixed (T* ptr = vertices)
            {
                AstralCanvas.VertexBuffer_SetData(handle, (IntPtr)ptr, (UIntPtr)vertices.Length);
            }
        }
        public void Dispose()
        {
            if (handle != IntPtr.Zero)
            {
                AstralCanvas.VertexBuffer_Deinit(handle);
            }
            handle = IntPtr.Zero;
        }
    }
}
